File: js/mostWater.js
/**
 * !DONE
 * @param {number[]} height
 * @return {number}
 * Given n non-negative integers a1, a2, ..., an , where each represents a point at coordinate (i, ai). n vertical lines are drawn such that the two endpoints of the line i is at (i, ai) and (i, 0). Find two lines, which, together with the x-axis forms a container, such that the container contains the most water.
 */
const mostWater1 = (height) => {
  let currentMax = 0
  // brute force
  // loop through the array twice, 
  // while test the min height time the length vs the current max 
  for (let i = 0; i < height.length; i++) {
    for (let j = i + 1; j < height.length; j++) {

      const horz = j - i
      const vert = Math.min(height[i], height[j])
      const area = horz * vert
      area > currentMax ? currentMax = area : null
      console.log({horz: horz, vert: vert, area: area, currentMax: currentMax})
      
    }
  }
  return currentMax
}

const mostWater2 = height => {
  // Start at both ends and move towards the mioddle
  let maxWater = 0
  let i = 0
  let j = height.length - 1
  while (i < j) {
    let len = j - i
    let minHeight = Math.min(height[i], height[j])
    let area = len * minHeight
    maxWater = Math.max(maxWater, area)
    height[i] < height[j] ? i++ : j--
  }
  return maxWater
}

// Testing

let height = [1, 8, 6, 2, 5, 4, 8, 3, 7]
let out = 49
let answer = mostWater2(height)
console.log(answer, out, answer === out)

File: python/most_water.py
""" 
You are given an integer array height of length n. There are n vertical lines drawn such that the two endpoints of the ith line are (i, 0) and (i, height[i]).

Find two lines that together with the x-axis form a container, such that the container contains the most water.

Return the maximum amount of water a container can store.

Notice that you may not slant the container.

 """

from ipdb import set_trace
from py_term_helpers import top_wrap, kv_print


def max_area(heights):
    # Two pointer
    # start with current max and right/left pointers
    # loop until the pointers cross
    # for each
    #   calc the area by:
    #       width = right - left
    #       height = min(height[right], height[left])
    #
    cur_max, l, r = 0, 0, len(heights) - 1
    while l < r:
        width = r - l
        height = min(heights[l], heights[r])
        cur_area = width * height
        cur_max = max(cur_area, cur_max)

        if heights[r] >= heights[l]:
            l += 1
        else:
            r -= 1
        kv_print((width, height, cur_area, cur_max))
    print(cur_max)
    return cur_max


top_wrap("MOST WATER")

height1 = [1, 8, 6, 2, 5, 4, 8, 3, 7]
# Output: 49

height2 = [1, 1]
# Output: 1

max_area(height1)
max_area(height2)
